ros2doctor/ros2doctor/api/package.py

```diff
@@ -0,0 +1,126 @@
+    distro_name = distro_name.lower()
+    url = rosdistro.get_index_url()
+    i = rosdistro.get_index(url)
+    distro_data = rosdistro.get_distribution(i, distro_name).get_data()
+    return distro_data.get('repositories')
+
+
+def get_packages_versions(distro_repos_info: dict) -> dict:
+    """
+    Return all package names and corresponding versions in rosdistro.
+
+    :param: a dictionary of repo info
+    :return: a dictionary contains package names and versions
+    """
+    packages_versions = {}
+    for _, info in distro_repos_info.items():
+        release = info.get('release')
+        if release:
+            version = release.get('version')
+            packages = release.get('packages')
+            if packages:
+                for p in packages:
+                    packages_versions[p] = version
+    return packages_versions
+
+
+class PackageCheck(DoctorCheck):
+    """Check local package versions against release versions on rosdistro."""
+
+    def category(self):
+        return 'package'
+
+    def check(self):
+        """Check packages within the directory where command is called."""
+        result = Result()
+        try:
+            distro_repos_info = get_ros_repos_info()
+            packages_versions = get_packages_versions(distro_repos_info)
+        except (AttributeError, RuntimeError):
+            result.add_error('ERROR: Unable to obtain current distro package information \
+                from rosdistro.')
+            return result
+        local_packages_prefixes = get_packages_with_prefixes()
+        if not local_packages_prefixes or not packages_versions:
+            result.add_error('ERROR: local packages info or distro packages info is empty.')
+            return result
+        for package_name, package_prefix in local_packages_prefixes.items():
+            file_path = os.path.join(package_prefix, 'share', package_name)
+            required_ver = packages_versions.get(package_name)
+            try:
+                package_obj = parse_package(file_path)
+                local_ver = package_obj.version
+            except (AttributeError, IOError, InvalidPackage):
+                result.add_warning('Unable to parse `%s` package.xml file.' % package_name)
+                local_ver = ''
+            if required_ver and local_ver:
+                if version.parse(local_ver).base_version < version.parse(required_ver).base_version:
+                    result.add_warning('%s local version %s does not match required version %s.'
+                                       % (package_name, package_obj.version, required_ver))
+        return result
+
+
+class PackageReport(DoctorReport):
+    """Report local package versions and release versions on rosdistro."""
+
+    def category(self):
+        return 'package'
+
+    def report(self):
+        """Report packages within the directory where command is called."""
+        report = Report('PACKAGE VERSIONS')
+        try:
```

Comment:
It looks like the bulk of this logic is identical to PackageCheck logic. It would be better to de-duplicate by moving it to a common function (or two) for reuse.